features/steps/validation_handling.py

```diff
@@ -116,16 +116,20 @@ def handle_then(context, fn, **kwargs):
     # if 'instances' are not actual ifcopenshell.entity_instance objects, but e.g. tuple of string values then get the actual instances from the stack tree
     activation_instances = misc.do_try(lambda: get_stack_tree(context)[-1], instances)
 
-    validation_outcome = ValidationOutcome(
-        outcome_code=ValidationOutcomeCode.EXECUTED,  # "Executed", but not no error/pass/warning #deactivated for now
-        observed=None,
-        expected=None,
-        feature=context.feature.name,
-        feature_version=misc.define_feature_version(context),
-        severity=OutcomeSeverity.EXECUTED,
-        validation_task_id=context.validation_task_id
-    )
-    context.gherkin_outcomes.append(validation_outcome)
+    #ensure the rule is not activated when there are no instances
+    #in case there are no instances but the rule is applicable (e.g. SPS001), then the rule is still activated and will return either a pass or an error
+    if misc.do_try(lambda: (lambda f: f(f))(lambda f: lambda data: bool(data) and (not isinstance(data, (list, tuple)) or any(f(f)(item) for item in data)))(context.instances), context.applicable):
```

Comment:
Sorry, but this really above my head.

`lambda: (lambda f: f(f))(lambda f: lambda data:` ... `f(f)(item)`

Can you explain what this does and then we try to come up with a simpler way to express this.

Reply:
It's a recursive lambda function to check whether the list contains meaningful entities. See also[ Fixed-point combinator](https://en.wikipedia.org/wiki/Fixed-point_combinator) (also includes an example in C++).  I wrote it down simpler before and was wondering if it was possible to write recursive functions without actually defining one (so with lambda instead). Turns out it is. But I agree, it's awful to read ..
The simpler version would be; 
```
def contains_truthy(data):
    if not data:
        return False
    if isinstance(data, (list, tuple)):
        return any(contains_truthy(item) for item in data)
    return True

result = misc.do_try(lambda: contains_truthy(context.instances), context.applicable)
```
The `misc.do_try` part is in case we don't have any context.instances instantiated. In this case, we look for applicability (e.g. correct schema). As a concrete example, (a part of) SPS001;
```
    Given A file with Model View Definition "CoordinationView"
    Given A file with Schema "IFC2X3"

    Then There must be at most 1 instance(s) of IfcSite
```
We don't have any Given statement that selects instances from the model (e.g. Given an IfcSite). The activation criteria here is therefore not related to instances but merely to the schema identifier/MVD. We discussed last Thursday that this could be desirable, but I suppose it's not set in stone. The alternative would be to re-write rules so that there are always activating instances. For this rule, it would mean that 0 instances of IfcSite would mean 'not-activated' and more than 1 IfcSite would be an error. In Gherkin, it will look something like; 
```
Given a correct Schema & MVD 
Given An IfcSite (or 'All instances of IfcSite', similar as GRF001) 
Then the length must be 1 at depth 1
```
Another approach would be to create a '000' rule for each functional part. This way, we clearly define what we mean when we say the software supports a feature. Then, we can use this rule to determine whether a block should be colored grey or green in case the validation passes.

Reply:
Very nice, didn't spot the y-combinator immediately, should have, but it's not very pythonic because we're not actually doing lambda calculus. Isn't it the similar to e.g `any(utils.recursively_flatten(context.instances))`? Which I would say is a bit more direct way of writing this.

https://github.com/buildingSMART/ifc-gherkin-rules/blob/f7dcaac762beb0c613d46f29fe94ca8aa778f9dd/features/steps/thens/existance.py#L52

---

> The misc.do_try part is in case we don't have any context.instances instantiated

I would prefer if we not catch arbitrary exceptions for known situations. `getattr(context.instances, [])` would work equally well in that case (?) and doesn't result in silencing arbitrary exceptions that we might actually want to know about.

---

To repeat I like what you did here, but since this involves the formalization of our certification logic I prefer we write it in a way that a typical MBA can almost comprehend this. Lambda calc and y combinators is really a bit too much I think.

---

Final note. In python you can actually have recursive lambdas by just naming them.

```
>>> f = lambda n: 0 if n == 1 else 1 if n == 2 else f(n-2) + f(n-1)
>>>
>>> f(10)
34
```

Reply:
Thanks for the explanations :-) 

> Isn't it the similar to e.g any(utils.recursively_flatten(context.instances))

Yes. I think I was a bit too enthusiastic about the lambda recursive lambdas and overlooked the more simple solution. 

> getattr(context.instances, []) would work equally well in that case (?) 

I think it could be even simpler. In a couple of lines above, we already define `instance` as
```
instances = getattr(context, 'instances', None) or (context.model.by_type(kwargs.get('entity')) if 'entity' in kwargs else [])
```
Knowing this, we can simply do; 
```
is_activated = any(misc.recursive_flatten(instances)) if instances else context.applicable
    if is_activated:
       # pass to db
```
Which is a lot more comprehensible for a typical MBA. 

---

> Final note. In python you can actually have recursive lambdas by just naming them.

Ah, I thought that lambda functions could not reference themselves. I was also looking to immediately invoke the function. Not really with a y combinator, but an outer lambda function. 
```
>>> (lambda f, n: f(f, n))(lambda f, n: 0 if n == 1 else 1 if n == 2 else f(f, n-2) + f(f, n-1), 10)
34
```